src/main/haskell/language-kore/src/Data/Kore/Unification/UnificationAlgorithm.hs

```diff
@@ -0,0 +1,264 @@
+              Pure patterns
+Copyright   : (c) Runtime Verification, 2018
+License     : UIUC/NCSA
+Maintainer  : phillip.harris@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE Rank2Types             #-}
+{-# LANGUAGE AllowAmbiguousTypes    #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE DeriveFunctor          #-}
+{-# LANGUAGE DeriveTraversable      #-}
+{-# LANGUAGE PatternSynonyms        #-}
+{-# LANGUAGE TemplateHaskell        #-}
+{-# LANGUAGE ConstraintKinds        #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE StandaloneDeriving     #-}
+{-# LANGUAGE BangPatterns           #-}
+
+
+module Data.Kore.Unification.UnificationAlgorithm where
+
+import qualified Data.Set as S
+import qualified Data.Map.Strict as M
+import           Data.Fix
+
+import           Control.Lens
+import           Control.Lens.Operators
+import           Control.Monad.State
+import           Control.Monad.Reader
+import           Control.Monad.Except
+
+import           Data.Kore.AST.Common
+import           Data.Kore.AST.Kore
+import           Data.Kore.AST.PureML
+import           Data.Kore.AST.MetaOrObject
+import           Data.Kore.IndexedModule.MetadataTools
+
+import           Data.Kore.Proof.ProofSystemWithHypos
+import           Data.Kore.Unification.UnificationRules
+import           Data.Kore.Unparser.Unparse
+
+import Debug.Trace
+
+{- 
+NOTE:
+1) Carrying around ix=Int and constantly dereferencing
+is too cumbersome. Instead: carry around the proof lines,
+and just hashcons when constructing new Rule ix?
+2) Handling multiple states with lenses is not pretty,
+but with a few combinators it becomes fairly unobtrusive
+and I don't see many better options. 
+-}
+
+data UnificationState level
+  = UnificationState 
+  { _activeSet :: ![Idx]
+  , _finishedSet :: ![Idx]
+  , _proof :: !(Proof Int (UnificationRules level) (Term level))
+  } 
+  deriving(Show)
+
+data UnificationError level
+    = ConstructorClash     (Term level) (Term level)
+    | NonConstructorHead   (Term level)
+    | NonFunctionalPattern (Term level)
+    | OccursCheck          (Term level) (Term level)
+  deriving (Eq, Show)
+
+makeLenses ''UnificationState
+
+type UnificationContext level m = 
+  ( MonadState (UnificationState level) m
+  , MonadReader (MetadataTools level) m
+  , MonadError (UnificationError level) m
+  ) 
+
+type Unification level a = 
+  ReaderT (MetadataTools level) (
+  StateT (UnificationState level) (
+  ExceptT (UnificationError level) 
+  Identity))
+  a
+
+unificationProcedure 
+  :: (MetaOrObject level
+  ,  UnificationContext level m)
+  => Term level
+  -> Term level
+  -> m Idx
+unificationProcedure a b = do
+  (ixMGU, forwardsDirection) <- proveForwardsDirection a b
+  backwardsDirection <-
+    (claim <$> proof %%%= lookupLine ixMGU)
```

Comment:
What would be the standard way to search for the meaning of `%%%=`? I thought about the following options:

1. Searching through the pull request only finds usages of this operator. 
2. Hoogle - does not find anything
3. Lens' cheat sheet does not have it, and it's not in the [obvious place](https://hackage.haskell.org/package/lens-4.16.1/docs/Control-Lens-Operators.html) given the imports (Lens is the package I know least).
4. A Google search does not find it.

Could you use explicit imports or qualified imports?
https://docs.google.com/document/d/1h_rNcOzZQi0pCKMCdtmUm-MDKklY33LujpMzCEuQPCQ/edit#heading=h.b53jtytkm8xf